lib/integrations/awsoidc/deploydbservice.go

```diff
@@ -0,0 +1,655 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/service/ecs"
+	ecsTypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v3"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/config"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+var (
+	// launcTypeFargateString is the FARGATE LaunchType converted into a string.
+	launcTypeFargateString = string(ecsTypes.LaunchTypeFargate)
+	// requiredCapacityProviders contains the FARGATE type which is required to deploy a DatabaseService.
+	requiredCapacityProviders = []string{launcTypeFargateString}
+
+	// Ensure Cpu and Memory use one of the allowed combinations:
+	// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html
+	taskCPU = "256"
+	taskMem = "512"
+
+	// taskAgentContainerName is the name of the container to run within the Task.
+	// Each task supports multiple containers, but, currently, there's only one being used.
+	// This is also going to be the NodeName of the teleport instance.
+	taskAgentContainerName = "ecs-discovery-dbservice"
+
+	// oneAgent is used to define the desired agent count when creating a service.
+	oneAgent = int32(1)
+
+	// serviceForceDeletion indicates that the service must be deleted even if it has running tasks.
+	serviceForceDeletion = true
+)
+
+const (
+	// teleportContainerImageFmt is the Teleport Container Image to be used
+	teleportContainerImageFmt = "public.ecr.aws/gravitational/teleport-distroless:%s"
+
+	// clusterStatusActive is the string representing an ACTIVE ECS Cluster.
+	clusterStatusActive = "ACTIVE"
+	// clusterStatusInactive is the string representing an INACTIVE ECS Cluster.
+	clusterStatusInactive = "INACTIVE"
+
+	// serviceStatusActive is the string representing an ACTIVE ECS Service.
+	serviceStatusActive = "ACTIVE"
+	// serviceStatusDraining is the string representing an DRAINING ECS Service.
+	serviceStatusDraining = "DRAINING"
+)
+
+// DeployDBServiceRequest contains the required fields to deploy a Database and a Discovery Service.
+type DeployDBServiceRequest struct {
```

Comment:
Can we keep the API not db-specific? Can it just accept a config to start the agent with so it can be extended to other types of agents?